File: objc/objc/QLPageView.m
//
//  QLPageView.m
//  objc
//
//  Created by Qingwei Lan on 4/6/16.
//  Copyright © 2016 Qingwei Lan. All rights reserved.
//

#import "QLPageView.h"

@interface QLPageView () <UIScrollViewDelegate>

// Currently selected page index
@property (nonatomic, readwrite) NSInteger selectedIndex;

// Scroll view containing the views of the pages
@property (nonatomic, strong) UIScrollView *containerScrollView;
// View containing the buttons of the pages
@property (nonatomic, strong) UIView *buttonBar;
// Circle used to indicate selected page
@property (nonatomic, strong) UIView *selectionIndicator;

// Array that keeps hold of all buttons
@property (nonatomic, strong) NSArray *buttons;
// Array that keeps hold of all labels
@property (nonatomic, strong) NSArray *labels;

@end

@implementation QLPageView
{
    CGFloat gapV;
    CGFloat gapH;
    CGFloat selectionIndicatorY;
    CGFloat buttonWidth;
    CGFloat labelWidth;
    
    struct {
        bool initialIndexForPageInPageView;
        bool numberOfPagesInPageView;
        bool titleForLabelForPageAtIndex;
        bool selectedIndexForPageView;
    } dataSourceCan;
    
    struct {
        bool heightForButtonBarForPageView;
        bool colorForButtonBarForPageView;
        bool colorForButtonBarSelectionIndicatorForPageView;
        bool onTintColorForControlSwitchForPageView;
        bool fontForButtonsForPageView;
        bool fontForLabelsForPageView;
        bool controlSwitchDidChangeValue;
        bool didMoveToPage;
    } delegateCan;
}

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    
    if (self) {
        _pageViewStyle = QLPageViewButtonBarStyleDefault;
        _buttonBarHeight = 50.0;
        _buttonBarColor = [UIColor colorWithRed:25.0/255 green:110.0/255 blue:180.0/255 alpha:1];
        _switchOnTintColor = [UIColor yellowColor];
        _buttonBarSelectionIndicatorColor = [UIColor whiteColor];
        _buttonFont = [UIFont fontWithName:@"Helvetica-Bold" size:14.0];
        _labelFont = [UIFont fontWithName:@"Helvetica" size:10.0];
    }
    
    return self;
}

- (void)didMoveToSuperview
{
    [super didMoveToSuperview];
    
    if (self.superview) {
        [self reloadPageView];
    }
}

- (void)reloadPageView
{
    if (!self.dataSource) {
        return;
    }
    
    NSInteger selectedIndex = self.selectedIndex;
    if (dataSourceCan.selectedIndexForPageView) {
        selectedIndex = [self.dataSource selectedIndexForPageView:self];
    }
    
    // Remove all subviews
    [self.subviews enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        [obj removeFromSuperview];
    }];
    
    [self loadContainerScrollView];
    [self loadButtonBar];
    
    [self gotoPage:selectedIndex animated:NO];
    [self selectButtonAtIndex:selectedIndex];
}

- (void)setDataSource:(id<QLPageViewDataSource>)dataSource
{
    _dataSource = dataSource;
    
    dataSourceCan.initialIndexForPageInPageView = [self.dataSource respondsToSelector:@selector(initialIndexForPageInPageView:)];
    dataSourceCan.numberOfPagesInPageView = [self.dataSource respondsToSelector:@selector(numberOfPagesInPageView:)];
    dataSourceCan.selectedIndexForPageView = [self.dataSource respondsToSelector:@selector(selectedIndexForPageView:)];
    dataSourceCan.titleForLabelForPageAtIndex = [self.dataSource respondsToSelector:@selector(pageView:titleForLabelForPageAtIndex:)];
}

- (void)setDelegate:(id<QLPageViewDelegate>)delegate
{
    delegateCan.colorForButtonBarForPageView = [self.delegate respondsToSelector:@selector(colorForButtonBarForPageView:)];
    delegateCan.colorForButtonBarSelectionIndicatorForPageView = [self.delegate respondsToSelector:@selector(colorForButtonBarSelectionIndicatorForPageView:)];
    delegateCan.onTintColorForControlSwitchForPageView = [self.delegate respondsToSelector:@selector(onTintColorForControlSwitchForPageView:)];
    delegateCan.controlSwitchDidChangeValue = [self.delegate respondsToSelector:@selector(pageView:controlSwitchDidChangeValue:)];
    delegateCan.didMoveToPage = [self.delegate respondsToSelector:@selector(pageView:didMoveToPage:)];
    delegateCan.fontForButtonsForPageView = [self.delegate respondsToSelector:@selector(fontForButtonsForPageView:)];
    delegateCan.fontForLabelsForPageView = [self.delegate respondsToSelector:@selector(fontForLabelsForPageView:)];
    delegateCan.heightForButtonBarForPageView = [self.delegate respondsToSelector:@selector(heightForButtonBarForPageView:)];
}

#pragma mark - Container Scroll View Methods

- (void)loadContainerScrollView
{
    CGFloat buttonBarHeight = self.buttonBarHeight;
    if (delegateCan.heightForButtonBarForPageView) {
        buttonBarHeight = [self.delegate heightForButtonBarForPageView:self];
    }
    NSInteger numberOfPages = self.numberOfPages;
    if (dataSourceCan.numberOfPagesInPageView) {
        numberOfPages = [self.dataSource numberOfPagesInPageView:self];
    }
    
    // Initialize container scroll view
    _containerScrollView = [[UIScrollView alloc] initWithFrame:CGRectMake(0, buttonBarHeight, self.frame.size.width, self.frame.size.height - buttonBarHeight)];
    
    [self addSubview:self.containerScrollView];
    
    // Setting the height to 1 will allow only horizontal scrolls
    self.containerScrollView.contentSize = CGSizeMake(CGRectGetWidth(self.frame) * numberOfPages, 1);
    
    self.containerScrollView.delegate = self;
    self.containerScrollView.pagingEnabled = YES;
    self.containerScrollView.showsHorizontalScrollIndicator = NO;
    self.containerScrollView.showsVerticalScrollIndicator = NO;
    self.containerScrollView.directionalLockEnabled = YES;
    self.containerScrollView.bounces = NO;
    
    for (NSInteger i = 0; i < numberOfPages ; i++) {
        [self loadPageWithIndex:i];
    }
}

- (void)loadPageWithIndex:(NSInteger)index
{
    // Get the view from the dataSource
    UIView *page = [self.dataSource pageView:self viewForPageAtIndex:index];
    
    // Setup the frame of the page
    CGRect frame = self.containerScrollView.frame;
    frame.origin.x = frame.size.width * index;
    frame.origin.y = 0;
    page.frame = frame;
    
    [self.containerScrollView addSubview:page];
}

#pragma mark - Button Bar Methods

- (void)loadButtonBar
{
    CGFloat buttonBarHeight = self.buttonBarHeight;
    if (delegateCan.heightForButtonBarForPageView) {
        buttonBarHeight = [self.delegate heightForButtonBarForPageView:self];
    }
    UIColor *buttonBarColor = self.buttonBarColor;
    if (delegateCan.colorForButtonBarForPageView) {
        buttonBarColor = [self.delegate colorForButtonBarForPageView:self];
    }
    
    
    // Initialize button bar
    _buttonBar = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.frame.size.width, buttonBarHeight)];
    self.buttonBar.backgroundColor = buttonBarColor;
    [self addSubview:self.buttonBar];
    
    [self loadButtons];
}

- (void)loadButtons
{
    NSInteger numberOfPages = self.numberOfPages;
    if (dataSourceCan.numberOfPagesInPageView) {
        numberOfPages = [self.dataSource numberOfPagesInPageView:self];
    }
    UIColor *selectionIndicatorColor = self.buttonBarSelectionIndicatorColor;
    if (delegateCan.colorForButtonBarSelectionIndicatorForPageView) {
        selectionIndicatorColor = [self.delegate colorForButtonBarSelectionIndicatorForPageView:self];
    }
    UIColor *buttonBarColor = self.buttonBarColor;
    if (delegateCan.colorForButtonBarForPageView) {
        buttonBarColor = [self.delegate colorForButtonBarForPageView:self];
    }
    
    NSMutableArray *buttons = [[NSMutableArray alloc] initWithCapacity:0];
    NSMutableArray *labels = [[NSMutableArray alloc] initWithCapacity:0];
    
    switch (self.pageViewStyle)
    {
        case QLPageViewButtonBarStyleDefault:
        {
            gapV = self.buttonBar.bounds.size.height/6.0;
            buttonWidth = 4.0*gapV;
            gapH = (self.buttonBar.bounds.size.width - numberOfPages*buttonWidth - (((numberOfPages-1) < 0) ? 0 : (numberOfPages-1))*2*gapV)/2.0;
            selectionIndicatorY = gapV;
            
            break;
        }
            
        case QLPageViewButtonBarStyleWithLabel:
        {
            
            gapV = self.buttonBar.bounds.size.height/7.0;
            buttonWidth = 4.0*gapV;
            labelWidth = 1.5*gapV;
            gapH = (self.buttonBar.bounds.size.width - numberOfPages*buttonWidth - (((numberOfPages-1) < 0) ? 0 : (numberOfPages-1))*2*gapV)/2.0;
            selectionIndicatorY = 2*gapV;
            
            break;
        }
            
        case QLPageViewButtonBarStyleWithRightSwitch:
        {
            UIColor *onTintColor = self.switchOnTintColor;
            if (delegateCan.onTintColorForControlSwitchForPageView) {
                onTintColor = [self.delegate onTintColorForControlSwitchForPageView:self];
            }
            CGFloat switchX = self.buttonBar.bounds.size.width*(3/4.0);
            CGFloat switchY = self.buttonBar.bounds.size.height/2.0 - 15.0;
            
            // Add switch
            _controlSwitch = [[UISwitch alloc] initWithFrame:CGRectMake(switchX, switchY, 10.0, 10.0)];
            [self.controlSwitch addTarget:self action:@selector(switchChangedValue:) forControlEvents:UIControlEventValueChanged];
            [self.buttonBar addSubview:self.controlSwitch];
            self.controlSwitch.onTintColor = onTintColor;
            
            CGFloat contentWidth = switchX;
            gapV = self.buttonBar.bounds.size.height/6.0;
            buttonWidth = 4.0*gapV;
            gapH = (contentWidth - numberOfPages*buttonWidth) / (numberOfPages+1);
            selectionIndicatorY = gapV;
            
            break;
        }
            
        default: break;
    }
    
    // Load selection indicator
    _selectionIndicator = [[UIView alloc] initWithFrame:CGRectMake(gapH, selectionIndicatorY, buttonWidth, buttonWidth)];
    self.selectionIndicator.backgroundColor = selectionIndicatorColor;
    self.selectionIndicator.layer.cornerRadius = 2*gapV;
    [self.buttonBar addSubview:self.selectionIndicator];
    
    // Load buttons & Labels
    for (NSInteger i = 0; i < numberOfPages; i++) {
        // Add buttons
        UIButton *button;
        switch (self.pageViewStyle)
        {
            case QLPageViewButtonBarStyleDefault:
            {
                button = [[UIButton alloc] initWithFrame:CGRectMake(gapH+i*6*gapV, 2*gapV, buttonWidth, buttonWidth)];
                break;
            }
                
            case QLPageViewButtonBarStyleWithLabel:
            {
                button = [[UIButton alloc] initWithFrame:CGRectMake(gapH+i*6*gapV, 2*gapV, buttonWidth, buttonWidth)];
                
                // Add labels
                UILabel *label = [[UILabel alloc] initWithFrame:CGRectMake(gapH+i*6*gapV, 0, buttonWidth, labelWidth)];
                if (dataSourceCan.titleForLabelForPageAtIndex) {
                    label.text = [self.dataSource pageView:self titleForLabelForPageAtIndex:i];
                } else {
                    label.text = @"";
                }
                label.font = self.labelFont;
                label.textColor = selectionIndicatorColor;
                label.textAlignment = NSTextAlignmentCenter;
                label.userInteractionEnabled = NO;
                [self.buttonBar addSubview:label];
                [labels addObject:label];
                break;
            }
                
            case QLPageViewButtonBarStyleWithRightSwitch:
            {
                button = [[UIButton alloc] initWithFrame:CGRectMake(gapH+i*(gapH+buttonWidth), gapV, buttonWidth, buttonWidth)];
                break;
            }
                
            default: break;
        }
        
        button.layer.cornerRadius = 2*gapV;
        button.backgroundColor = [UIColor clearColor];
        [button setTitle:[self.dataSource pageView:self titleForButtonForPageAtIndex:i] forState:UIControlStateNormal];
        [button setTitleColor:selectionIndicatorColor forState:UIControlStateNormal];
        button.titleLabel.font = self.buttonFont;
        button.titleLabel.textAlignment = NSTextAlignmentCenter;
        [button addTarget:self action:@selector(selectPage:) forControlEvents:UIControlEventTouchUpInside];
        [self.buttonBar addSubview:button];
        [buttons addObject:button];
    }
    
    _buttons = buttons;
    _labels = labels;
}

#pragma mark - Actions

- (void)selectPage:(UIButton *)sender
{
    NSInteger index = [self.buttons indexOfObject:sender];
    [self gotoPage:index animated:YES];
}

- (void)selectButtonAtIndex:(NSInteger)index
{
    UIColor *buttonBarColor = self.buttonBarColor;
    if (delegateCan.colorForButtonBarForPageView) {
        buttonBarColor = [self.delegate colorForButtonBarForPageView:self];
    }
    if ([self.buttons count]) {
        UIButton *button = [self.buttons objectAtIndex:index];
        [button setTitleColor:buttonBarColor forState:UIControlStateNormal];
    }
}

- (void)deselectButtonAtIndex:(NSInteger)index
{
    UIColor *selectionIndicatorColor = self.buttonBarSelectionIndicatorColor;
    if (delegateCan.colorForButtonBarSelectionIndicatorForPageView) {
        selectionIndicatorColor = [self.delegate colorForButtonBarSelectionIndicatorForPageView:self];
    }
    if ([self.buttons count]) {
        UIButton *button = [self.buttons objectAtIndex:index];
        [button setTitleColor:selectionIndicatorColor forState:UIControlStateNormal];
    }
}

- (void)setSelectedIndex:(NSInteger)selectedIndex
{
    [self deselectButtonAtIndex:self.selectedIndex];
    _selectedIndex = selectedIndex;
    [self selectButtonAtIndex:self.selectedIndex];
}

- (void)gotoPage:(NSInteger)index animated:(BOOL)animated
{
    CGRect bounds = self.containerScrollView.bounds;
    bounds.origin.x = CGRectGetWidth(bounds) * index;
    bounds.origin.y = 0;
    [self.containerScrollView scrollRectToVisible:bounds animated:animated];
}

- (void)switchChangedValue:(UISwitch *)sender
{
    if (delegateCan.controlSwitchDidChangeValue) {
        [self.delegate pageView:self controlSwitchDidChangeValue:sender.isOn];
    }
}

#pragma mark - UIScrollViewDelegate

- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    CGFloat pageWidth = CGRectGetWidth(self.containerScrollView.frame);
    
    CGFloat contentWidth = self.containerScrollView.contentSize.width - pageWidth;
    
    if (contentWidth != 0) {
        CGFloat scrolledPercentage = self.containerScrollView.contentOffset.x / contentWidth;
        UIButton *lastButton = self.buttons.lastObject;
        UIButton *firstButton = self.buttons.firstObject;
        CGFloat barWidth = lastButton.frame.origin.x - firstButton.frame.origin.x;
        
        CGRect frame = self.selectionIndicator.frame;
        frame.origin.x = firstButton.frame.origin.x + scrolledPercentage * barWidth;
        self.selectionIndicator.frame = frame;
        
        NSInteger index = floor((self.containerScrollView.contentOffset.x - pageWidth / 2) / pageWidth) + 1;
        if (index != self.selectedIndex) {
            self.selectedIndex = index;
            if (delegateCan.didMoveToPage) {
                [self.delegate pageView:self didMoveToPage:self.selectedIndex];
            }
        }
    }
}

@end

File: swift/page/QLPageView.swift
//
//  QLPageView.swift
//  page
//
//  Created by Qingwei Lan on 4/9/16.
//  Copyright © 2016 Qingwei Lan. All rights reserved.
//

import Foundation
import UIKit

enum ButtonBarStyle {
    case Default
    case WithLabel
    case WithRightSwitch
}

// ------------------------------------------------------
// This protocol represents the data model object
// ------------------------------------------------------
@objc protocol QLPageViewDataSource: class {
    
    /**
     Title of the button associated with the page at index.
     */
    func titleForButtonForPageAtIndex(sender: QLPageView, index: Int) -> String
    
    /**
     View of the page at index.
     */
    func viewForPageAtIndex(sender: QLPageView, index: Int) -> UIView
    
    /**
     The page where the page view starts, default is 0.
     The initial displayed page always has index 0.
     Index must be in range: 0 <= index < numberOfPages.
     */
    optional func initialIndexForPageInPageView(sender: QLPageView) -> Int
    
    /**
     The number of pages in the page view, default is 0.
     Maximum number of pages is 6.
     */
    optional func numberOfPagesInPageView(sender: QLPageView) -> Int
    
    /**
     Title of the button associated with the page at index.
     Only works if pageViewStyle is QLPageViewButtonStyleWithLabel.
     Default is nil.
     */
    optional func titleForLabelForPageAtIndex(sender: QLPageView, index: Int) -> String
    
    /**
     Selected index for page view, default is initialIndex.
     */
    optional func selectedIndexForPageView(sender: QLPageView) -> Int
}

// ------------------------------------------------------
// This protocol represents the display and behaviour
// ------------------------------------------------------
@objc protocol QLPageViewDelegate: class {
    
    /**
     Sets the height for the button bar.
     Default height is 50.0.
     */
    optional func heightForButtonBarForPageView(sender: QLPageView) -> CGFloat
    
    /**
     Sets the color of the button bar.
     Default is black color.
     */
    optional func colorForButtonBarForPageView(sender: QLPageView) -> UIColor
    
    /**
     Sets the color of the button bar selection indicator.
     Default is white color.
     */
    optional func colorForButtonBarSelectionIndicatorForPageView(sender: QLPageView) -> UIColor
    
    /**
     Sets the on tint color of the control switch.
     Default is yellow color.
     */
    optional func onTintColorForControlSwitchForPageView(sender: QLPageView) -> UIColor
    
    /**
     Sets the font of the button.
     Default font is Helvetica-Bold with size 14.0.
     */
    optional func fontForButtonsForPageView(sender: QLPageView) -> UIFont
    
    /**
     Sets the font of the label.
     Default font is Helvetica with size 10.0.
     */
    optional func fontForLabelsForPageView(sender: QLPageView) -> UIFont
    
    /**
     Called when Control Switch changes value.
     */
    optional func controlSwitchDidChangeValue(sender: QLPageView, value: Bool)
    
    /**
     Called when page changed.
     */
    optional func didMoveToPage(sender: QLPageView, page: Int)
}

@objc class QLPageView: UIView, UIScrollViewDelegate {
    
    weak var dataSource: QLPageViewDataSource? = nil
    weak var delegate: QLPageViewDelegate? = nil
    
    /**
     The declared style of the pageView that defines the appearance of the button bar
     */
    var pageViewStyle: ButtonBarStyle = .Default
    
    /**
     Initial index of the page view
     */
    var initialIndex = 0
    
    /**
     Number of pages in the page view
     */
    var numberOfPages = 0
    
    /**
     Height of the button bar
     */
    var buttonBarHeight: CGFloat = 50.0
    
    /**
     The control switch defined in the button bar, nil if pageViewStyle is not
     .WithRightSwitch
     */
    private(set) var controlSwitch: UISwitch? = nil
    
    /**
     Currently selected page index.
     */
    private(set) var selectedIndex = 0 {
        didSet {
            deselectButtonAtIndex(oldValue)
            selectButtonAtIndex(selectedIndex)
        }
    }
    
    /**
     Color of button bar (also color of button titles).
     Default color is black.
     */
    var buttonBarColor: UIColor = UIColor.blackColor()
    
    /**
     On tint color of the control switch.
     Default color is yellow.
     */
    var switchOnTintColor = UIColor.yellowColor()
    
    /**
     Font of button bar buttons.
     Default font is Helvetica-Bold with size 14.0.
     */
    var buttonFont = UIFont(name: "Helvetica-Bold", size: 14.0)!
    
    /**
     Font of button bar labels (if the labels exist).
     Default font is Helvetica with size 10.0.
     Can be nil unless pageViewStyle is .WithLabel.
     */
    var labelFont = UIFont(name: "Helvetica", size: 10.0)!
    
    
    /**
     Scroll view containing the views of the pages.
     */
    private var containerScrollView: UIScrollView? = nil
    
    /**
     View containing the buttons of the pages
     */
    private var buttonBar: UIView? = nil
    
    /**
     Circle used to indicate selected page
     */
    private var selectionIndicator: UIView? = nil
    
    /**
     Color of button bar selection indicator.
     Default color is white.
     */
    private var buttonBarSelectionIndicatorColor = UIColor.whiteColor()
    
    /**
     Array that keeps hold of all buttons
     */
    private var buttons = [UIButton]()
    
    /**
     Array that keeps hold of all labels
     */
    private var labels = [UILabel]()

    
    override init(frame: CGRect) {
        super.init(frame: frame)
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override func didMoveToSuperview() {
        if self.superview != nil {
            reloadPageView()
        }
    }
    
    func reloadPageView() {
        if dataSource == nil {
            return
        }
        if let selectedIndex = dataSource?.selectedIndexForPageView?(self) {
            self.selectedIndex = selectedIndex
        }
        if let initialIndex = dataSource?.initialIndexForPageInPageView?(self) {
            self.initialIndex = initialIndex
        }
        
        selectedIndex = initialIndex
        
        // Remove all subviews
        for view in self.subviews {
            view.removeFromSuperview()
        }
        
        // Initializations
        loadContainerScrollView()
        loadButtonBar()
        
        gotoPage(selectedIndex, animated: false)
        selectButtonAtIndex(selectedIndex)
    }
    
    func loadContainerScrollView() {
        if let buttonBarHeight = delegate?.heightForButtonBarForPageView?(self) {
            self.buttonBarHeight = buttonBarHeight
        }
        if let numberOfPages = dataSource?.numberOfPagesInPageView?(self) {
            self.numberOfPages = numberOfPages
        }
        
        containerScrollView = UIScrollView(frame: CGRectMake(0, buttonBarHeight, self.frame.size.width, self.frame.size.height - buttonBarHeight))
        self.addSubview(containerScrollView!)
        containerScrollView?.contentSize = CGSizeMake(CGRectGetWidth(self.frame) * CGFloat(numberOfPages), 1)
        
        containerScrollView?.delegate = self
        containerScrollView?.pagingEnabled = true
        containerScrollView?.showsHorizontalScrollIndicator = false
        containerScrollView?.showsVerticalScrollIndicator = false
        containerScrollView?.directionalLockEnabled = true
        containerScrollView?.bounces = false
        
        for i in 0...numberOfPages-1 {
            loadPageWithIndex(i)
        }
    }
    
    func loadPageWithIndex(index: Int) {
        // Get view from dataSource
        let page = dataSource?.viewForPageAtIndex(self, index: index)
        
        // Setup frame of page
        var frame = containerScrollView?.frame
        frame?.origin.x = frame!.size.width * CGFloat(index)
        frame?.origin.y = 0
        page?.frame = frame!
        
        containerScrollView?.addSubview(page!)
    }
    
    func loadButtonBar() {
        if let buttonBarHeight = delegate?.heightForButtonBarForPageView?(self) {
            self.buttonBarHeight = buttonBarHeight
        }
        if let buttonBarColor = delegate?.colorForButtonBarForPageView?(self) {
            self.buttonBarColor = buttonBarColor
        }
        
        // Initialize button bar
        buttonBar = UIView(frame: CGRectMake(0, 0, self.frame.size.width, buttonBarHeight))
        buttonBar?.backgroundColor = buttonBarColor
        self.addSubview(buttonBar!)
        
        loadButtons()
    }
    
    func loadButtons() {
        if let numberOfPages = dataSource?.numberOfPagesInPageView?(self) {
            self.numberOfPages = numberOfPages
        }
        if let buttonBarSelectionIndicatorColor = delegate?.colorForButtonBarSelectionIndicatorForPageView?(self) {
            self.buttonBarSelectionIndicatorColor = buttonBarSelectionIndicatorColor
        }
        if let buttonBarColor = delegate?.colorForButtonBarForPageView?(self) {
            self.buttonBarColor = buttonBarColor
        }
        
        var gapV: CGFloat = 0.0
        var gapH: CGFloat = 0.0
        var selectionIndicatorY: CGFloat = 0.0
        var buttonWidth: CGFloat = 0.0
        var labelWidth: CGFloat = 0.0
        
        switch pageViewStyle {
        case .Default:
            gapV = buttonBar!.bounds.size.height/6.0
            buttonWidth = 4.0*gapV
            let temp = ((numberOfPages-1) < 0) ? 0 : (numberOfPages-1)
            gapH = (buttonBar!.bounds.size.width - CGFloat(numberOfPages)*buttonWidth - CGFloat(temp)*2*gapV)/2.0
            selectionIndicatorY = gapV
            
        case .WithLabel:
            gapV = buttonBar!.bounds.size.height/7.0
            buttonWidth = 4.0*gapV
            labelWidth = 1.5*gapV
            let temp = ((numberOfPages-1) < 0) ? 0 : (numberOfPages-1)
            gapH = (buttonBar!.bounds.size.width - CGFloat(numberOfPages)*buttonWidth - CGFloat(temp)*2*gapV)/2.0
            selectionIndicatorY = 2*gapV
        
        case .WithRightSwitch:
            if let switchOnTintColor = delegate?.onTintColorForControlSwitchForPageView?(self) {
                self.switchOnTintColor = switchOnTintColor
            }
            
            let switchX = buttonBar!.bounds.size.width*(3/4.0)
            let switchY = buttonBar!.bounds.size.height/2.0 - 15.0
            
            // Add switch
            controlSwitch = UISwitch(frame: CGRectMake(switchX, switchY, 10.0, 10.0))
            controlSwitch?.addTarget(self, action: #selector(QLPageView.switchChangedValue(_:)), forControlEvents: UIControlEvents.ValueChanged)
            controlSwitch?.onTintColor = switchOnTintColor
            buttonBar?.addSubview(controlSwitch!)
            
            let contentWidth: CGFloat = switchX;
            gapV = buttonBar!.bounds.size.height/6.0;
            buttonWidth = 4.0*gapV;
            gapH = (contentWidth - CGFloat(numberOfPages)*buttonWidth) / (CGFloat(numberOfPages)+1);
            selectionIndicatorY = gapV;
        }
        
        // load selection indicator
        selectionIndicator = UIView(frame: CGRectMake(gapH, selectionIndicatorY, buttonWidth, buttonWidth))
        selectionIndicator?.backgroundColor = buttonBarSelectionIndicatorColor
        selectionIndicator?.layer.cornerRadius = 2*gapV
        buttonBar?.addSubview(selectionIndicator!)
        
        // load buttons & Labels
        for i in 0...numberOfPages-1 {
            var button: UIButton? = nil
            switch pageViewStyle {
            case .Default:
                button = UIButton(frame: CGRectMake(gapH+CGFloat(i)*6*gapV, 2*gapV, buttonWidth, buttonWidth))
                
            case .WithLabel:
                button = UIButton(frame: CGRectMake(gapH+CGFloat(i)*6*gapV, 2*gapV, buttonWidth, buttonWidth))
                
                // add labels
                let label = UILabel(frame: CGRectMake(gapH+CGFloat(i)*6*gapV, 0, buttonWidth, labelWidth))
                if let title = dataSource?.titleForLabelForPageAtIndex?(self, index: i) {
                    label.text = title
                } else {
                    label.text = ""
                }
                label.font = labelFont
                label.textColor = buttonBarSelectionIndicatorColor
                label.textAlignment = NSTextAlignment.Center
                label.userInteractionEnabled = false
                buttonBar?.addSubview(label)
                labels.append(label)
            
            case .WithRightSwitch:
                button = UIButton(frame: CGRectMake(gapH+CGFloat(i)*(gapH+buttonWidth), gapV, buttonWidth, buttonWidth))
            }
            
            button?.layer.cornerRadius = 2*gapV
            button?.backgroundColor = UIColor.clearColor()
            button?.setTitle(dataSource?.titleForButtonForPageAtIndex(self, index: i), forState: UIControlState.Normal)
            button?.setTitleColor(buttonBarSelectionIndicatorColor, forState: UIControlState.Normal)
            button?.titleLabel?.font = buttonFont
            button?.titleLabel?.textAlignment = NSTextAlignment.Center
            button?.addTarget(self, action: #selector(QLPageView.selectPage(_:)), forControlEvents: UIControlEvents.TouchUpInside)
            buttonBar?.addSubview(button!)
            buttons.append(button!)
        }
    }
    
    func gotoPage(index: Int, animated: Bool) {
        var bounds = containerScrollView?.bounds
        bounds?.origin.x = CGRectGetWidth(bounds!) * CGFloat(index)
        bounds?.origin.y = 0.0
        containerScrollView?.scrollRectToVisible(bounds!, animated: animated)
    }
    
    func selectButtonAtIndex(index: Int) {
        if let buttonBarColor = delegate?.colorForButtonBarForPageView?(self) {
            self.buttonBarColor = buttonBarColor
        }
        
        if buttons.count != 0 {
            let button = buttons[index]
            button.setTitleColor(buttonBarColor, forState: UIControlState.Normal)
        }
    }
    
    func deselectButtonAtIndex(index: Int) {
        if let buttonBarSelectionIndicatorColor = delegate?.colorForButtonBarSelectionIndicatorForPageView?(self) {
            self.buttonBarSelectionIndicatorColor = buttonBarSelectionIndicatorColor
        }
        
        if buttons.count != 0 {
            let button = buttons[index]
            button.setTitleColor(buttonBarSelectionIndicatorColor, forState: UIControlState.Normal)
        }
    }
    
    @objc func switchChangedValue(sender: UISwitch) {
        delegate?.controlSwitchDidChangeValue?(self, value: sender.on)
    }
    
    @objc func selectPage(sender: UIButton) {
        let i = buttons.indexOf(sender)
        gotoPage(i!, animated: true)
    }
    
    
    func scrollViewDidScroll(scrollView: UIScrollView) {
        let pageWidth = CGRectGetWidth(containerScrollView!.frame)
        let contentWidth = containerScrollView!.contentSize.width - pageWidth
        
        if contentWidth != 0 {
            let scrolledPercentage: CGFloat = containerScrollView!.contentOffset.x / contentWidth
            let lastButton = buttons.last
            let firstButton = buttons.first
            let barWidth = lastButton!.frame.origin.x - firstButton!.frame.origin.x
            
            var frame = selectionIndicator?.frame
            frame?.origin.x = firstButton!.frame.origin.x + scrolledPercentage * barWidth
            selectionIndicator?.frame = frame!
            
            let index = Int(floor((containerScrollView!.contentOffset.x - pageWidth / 2) / pageWidth) + 1)
            if index != selectedIndex {
                selectedIndex = index
                delegate?.didMoveToPage?(self, page: selectedIndex)
            }
        }
    }
    
}
